pennylane_qiskit/qiskit_device2.py

```diff
@@ -0,0 +1,606 @@
+        """
+
+        transpile_sig = inspect.signature(transpile).parameters
+
+        transpile_args = {arg: kwargs[arg] for arg in transpile_sig if arg in kwargs}
+        transpile_args.pop("circuits", None)
+        transpile_args.pop("backend", None)
+
+        return transpile_args
+
+    def compile_circuits(self, circuits):
+        r"""Compiles multiple circuits one after the other.
+
+        Args:
+            circuits (list[QuantumCircuit]): the circuits to be compiled
+
+        Returns:
+             list[QuantumCircuit]: the list of compiled circuits
+        """
+        # Compile each circuit object
+        compiled_circuits = []
+        transpile_args = self.get_transpile_args(self._kwargs)
+
+        for circuit in circuits:
+            compiled_circ = transpile(circuit, backend=self.backend, **transpile_args)
+            compiled_circ.name = f"circ{len(compiled_circuits)}"
+            compiled_circuits.append(compiled_circ)
+
+        return compiled_circuits
+
+    # pylint: disable=unused-argument
+    def execute(
+        self,
+        circuits: QuantumTape_or_Batch,
+        execution_config: ExecutionConfig = DefaultExecutionConfig,
+    ) -> Result_or_ResultBatch:
+        session = self._session or Session(backend=self.backend)
+
+        if not self._use_primitives:
+            results = self._execute_runtime_service(circuits, session=session)
+            return results
+
+        results = []
+
+        if isinstance(circuits, (QuantumTape, QuantumScript)):
+            circuits = [circuits]
+
+        for circ in circuits:
+            if isinstance(circ.measurements[0], (ExpectationMP, VarianceMP)):
+                execute_fn = self._execute_estimator
+            elif isinstance(circ.measurements[0], ProbabilityMP):
+                execute_fn = self._execute_sampler
+            else:
+                execute_fn = self._execute_runtime_service
+            results.append(execute_fn(circ, session))
+
+        if self._session is None:
+            # if this was not a session set on the device, but just one created for this execution, close
+            session.close()
+
+        return results
+
+    def _execute_runtime_service(self, circuits, session):
+        """Execution using old runtime_service (can't use runtime sessions)"""
+
+        # update kwargs in case Options has been modified since last execution
+        self._update_kwargs()
+
+        # in case a single circuit is passed
+        if isinstance(circuits, (QuantumTape, QuantumScript)):
+            circuits = [circuits]
+
+        qcirc = [
+            circuit_to_qiskit(circ, self.num_wires, diagonalize=True, measure=True)
+            for circ in circuits
+        ]
+        compiled_circuits = self.compile_circuits(qcirc)
+
+        program_inputs = {"circuits": compiled_circuits, "shots": self.shots.total_shots}
+
+        for kwarg in self._kwargs:
+            program_inputs[kwarg] = self._kwargs.get(kwarg)
```

Comment:
any reason why you're using `get`? I figure it's already in there
```suggestion
        for kwarg, value in self._kwargs.items():
            program_inputs[kwarg] = value
```